### PR TITLE
ath79: add mikrotik subtarget

### PR DESCRIPTION
**Pull request by rogerpueyo**
This is a RFC for adding a new ath79/mikrotik subtarget and moving the two currently supported MikroTik devices there. The main rationale behind this is the need for:

 - CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y to property to manage the 4K-sized flash partitions
 - FEATURES += minor and CONFIG_MTD_SPLIT_MINOR_FW=y for the MiNor (MikroTik NOR) split MTD format

Issues (so far):
 - mikrotik/config-default has the symbols from ath79/generic + ath79/nand + ar71xx/mikrotik, removing those that didn't seem needed (e.g. drivers not used). Are all of the current ones needed or is any missing?
 - the NAND board (rb922) still has the BOARD_NAME variable (related: https://github.com/openwrt/openwrt/pull/2790)
 - filenames get a duplicate "mikrotik" string, for the subtarget and for the vendor (e.g., openwrt-ath79-mikrotik-mikrotik_routerboard-922uags-5hpacd-squashfs-sysupgrade.bin)

---

This commit creates the ath79/mikrotik subtarget in order to support
MikroTik devices based on Qualcomm Atheros MIPS SoCs.

MikroTik devices need a couple of specific features: the split MiNOR
firmware MTD format, which is not used by other devices, and the 4k
sector erase size on SPI NOR storage, which can not be added to the
ath79/generic and ath79/nand subtargets now.

Additionally, the commit moves the two MikroTik devices already in
the generic and nand subtargets to this new one.

Tested on the RB922 board and the wAP AC router.

### CODE REVIEW

**Comment by adrianschmutzler** [comment]
Since IIRC there was no response at all on this subject, let's just go on with this as the subtarget seems (unfortunately) the best solution for now.

I will have a closer look on the code later.

> filenames get a duplicate "mikrotik" string, for the subtarget and for the vendor (e.g., openwrt-ath79-mikrotik-mikrotik_routerboard-922uags-5hpacd-squashfs-sysupgrade.bin)

Yes, that's not nice, but I'd still keep it for consistency.

When we added the first mikrotik device to ath79, I had the idea to call the subtarget "minor" instead of "mikrotik", but I'm not sure yet whether this is really an improvement for the entire situation.

**Comment by adrianschmutzler** [comment]
@xback Can you check the symbols in new config-default for mikrotik target

**Comment by rogerpueyo** [comment]
Thanks, Adrian. :)

> When we added the first mikrotik device to ath79, I had the idea to call the subtarget "minor" instead of "mikrotik", but I'm not sure yet whether this is really an improvement for the entire situation.

Yup, mikrotik_mikrotik is not nice, but is consistent. "minor" makes sense since the subtarget has the MiNor-type MTD, but it might be mistakenly understood by something in between "generic" and "tiny", who knows.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	mikrotik,routerboard-922uags-5hpacd|\
+	mikrotik,routerboard-wap-g-5hact2hnd)
+		ucidef_set_interface_lan "eth0"
+		;;
+	*)
+		ucidef_set_interfaces_lan_wan "eth0" "eth1"
+		;;
+	esac
+}
+
```
Comment (adrianschmutzler) [inline target/linux/ath79/mikrotik/base-files/etc/board.d/02_network]: I'd actually add an empty ath79_setup_macs block here for convenience:

```
ath79_setup_macs()
{
	local board="$1"

	case "$board" in
	esac

	[ -n "$lan_mac" ] && ucidef_set_interface_macaddr "lan" $lan_mac
	[ -n "$wan_mac" ] && ucidef_set_interface_macaddr "wan" $wan_mac
	[ -n "$label_mac" ] && ucidef_set_label_macaddr $label_mac
}
```

File: target/linux/ath79/mikrotik/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+board_config_update
+board=$(board_name)
+ath79_setup_interfaces $board
```
Comment (adrianschmutzler) [inline target/linux/ath79/mikrotik/base-files/etc/board.d/02_network]: And "ath79_setup_macs $board" here.

File: target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+platform_do_upgrade() {
+	local board=$(board_name)
+
+	case "$board" in
+	mikrotik,routerboard-922uags-5hpacd)
+		platform_do_upgrade_mikrotik_nand "$1"
+		;;
+	*)
+		nand_do_upgrade "$1"
```
Comment (adrianschmutzler) [inline target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh]: That should be default_do_upgrade, as it's currently used for the device coming from generic target.

File: target/linux/ath79/mikrotik/profiles/00-default.mk [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+# Copyright (C) 2020 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/Default
+	NAME:=Default Profile (all drivers)
+	PACKAGES:= \
+		kmod-usb-ohci kmod-usb2 kmod-usb-ledtrig-usbport nand-utils
```
Comment (adrianschmutzler) [inline target/linux/ath79/mikrotik/profiles/00-default.mk]: I wonder whether we need these packages; particularly ledtrig-usbport startles me.
However, i actually newer build with Default Profile ...

**Comment by adrianschmutzler** [comment]
Looks good despite my comments and the config-default files where I would rely on @xback for judgement.

Your comment about the possible "minor" misunderstanding complements my other concerns, so let's just stay with mikrotik.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/02_network (hunk 1 above)
Comment (rogerpueyo) [reply target/linux/ath79/mikrotik/base-files/etc/board.d/02_network]: Fine, adding it.

File: target/linux/ath79/mikrotik/base-files/etc/board.d/02_network (hunk 2 above)
Comment (rogerpueyo) [reply target/linux/ath79/mikrotik/base-files/etc/board.d/02_network]: Sure.

File: target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh (hunk 3 above)
Comment (rogerpueyo) [reply target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh]: True. I think I didn't go from ath79/generic => ath79/mikrotik => ath79/mikrotik on the NOR device, so I didn't catch this one. Fixing it.

File: target/linux/ath79/mikrotik/profiles/00-default.mk (hunk 4 above)
Comment (rogerpueyo) [reply target/linux/ath79/mikrotik/profiles/00-default.mk]: I guess this profile is a reminiscence from the old times; when I've used it it has always lacked one packet or another (ath10k... whatever). I'm removing the ledtrig-usbport, it's not needed at all anywhere in the subtarget.

**Comment by adrianschmutzler** [comment]
I've decided to merge this now as it somehow blocks development for mikrotik on ath79.
If we need to adjust kernel config afterwards, that should be easy to do ...

**Comment by rogerpueyo** [comment]
Great!

**Comment by ch6574** [comment]
Should this PR now mean we get snapshots appearing in https://downloads.openwrt.org/snapshots/targets/ath79/mikrotik/ ?

**Comment by f00b4r0** [comment]
The subtarget isn't autobuilt yet. Given it seems there is a problem with the handling of the soft_config partition erasesize, I recommend it remains that way for a bit longer to prevent "average users" from risking to trash their flash. I'm trying to isolate this regression.

**Comment by f00b4r0** [comment]
A bit of extra info, I think I spotted part of the problem (re erasesize issues): http://git.openwrt.org/9261e7447

I'm still investigating, it would also seem that 4K_SECTORS_LIMIT is also broken in 4.19

For now the ramips and ar71xx mikrotik devices are unaffected since they still run 4.14, in which the bug(s) is/are absent

**Pull request closed without merging**